File: windowsApp/src/Rally.Core/SettingsStore.cs
using System.Security.Cryptography;
using System.Text.Json;

// Settings + secrets. Mirrors SettingsStore.swift / PreferencesManager keys.
// Prefs in %LocalAppData%/Rally/settings.json; token + Xtream password via
// DPAPI on Windows, restricted file elsewhere.
namespace Rally.Core;

public sealed class SettingsStore
{
    public const string DefaultAddon = "https://sports.highfly.to/manifest.json";
    public static readonly string[] DefaultSportsOrder =
        ["NFL", "NCAAF", "NBA", "NCAAB", "MLB", "NHL", "EPL", "La Liga", "Champions League", "Serie A"];

    private readonly string _dir;
    private readonly string _prefsPath;
    private readonly string _secretsPath;
    private Dictionary<string, JsonElement> _prefs = new();
    private Dictionary<string, string> _secrets = new();

    public SettingsStore(string? dir = null)
    {
        _dir = dir ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Rally");
        Directory.CreateDirectory(_dir);
        _prefsPath = Path.Combine(_dir, "settings.json");
        _secretsPath = Path.Combine(_dir, "secrets.dat");
        Load();
    }

    public IptvProvider IptvProvider
    {
        get => Enum.TryParse<IptvProvider>(Str("iptv_provider"), out var p) ? p : IptvProvider.Stalker;
        set => Set("iptv_provider", value.ToString());
    }
    public string PortalUrl
    {
        get => Str("portal_url") ?? "";
        set => Set("portal_url", UrlNormalizer.NormalizePortal(value));
    }
    public string XtreamServerUrl
    {
        get => Str("xtream_server_url") ?? "";
        set => Set("xtream_server_url", UrlNormalizer.NormalizeXtreamServer(value));
    }
    public string XtreamUsername
    {
        get => Str("xtream_username") ?? "";
        set => Set("xtream_username", value.Trim());
    }
    public string XtreamPassword
    {
        get => Secret("xtream_password");
        set => SetSecret("xtream_password", value);
    }
    public string AuthToken
    {
        get => Secret("auth_token");
        set => SetSecret("auth_token", value);
    }
    public string MacAddress
    {
        get
        {
            var saved = (Str("mac_address") ?? "").Trim();
            if (saved.Length > 0) return saved;
            var mac = "00:1A:79:" + string.Join(":", Enumerable.Range(0, 3).Select(_ => RandomNumberGenerator.GetInt32(256).ToString("X2")));
            Set("mac_address", mac);
            return mac;
        }
        set => Set("mac_address", value.Trim());
    }
    public string SerialNumber
    {
        get => Str("serial_number") ?? "";
        set => Set("serial_number", value);
    }
    public string DeviceId
    {
        get => Str("device_id") ?? "";
        set => Set("device_id", value);
    }
    public List<string> StremioAddonUrls
    {
        get
        {
            if (Arr("stremio_addon_urls") is List<string> list && list.Count > 0) return list;
            var legacy = (Str("stremio_addon_url") ?? "").Trim();
            if (legacy.Length > 0 && UrlNormalizer.NormalizeAddon(legacy) is string norm) return [norm];
            return [DefaultAddon];
        }
        set => Set("stremio_addon_urls", value.Select(u => UrlNormalizer.NormalizeAddon(u)).OfType<string>().Distinct().OrderBy(x => x).ToList());
    }
    public string ChannelCacheIdentity
    {
        get => Str("channel_cache_identity") ?? "";
        set => Set("channel_cache_identity", value);
    }
    public bool SetupComplete
    {
        get => Bool("setup_complete", false);
        set => Set("setup_complete", value);
    }
    public bool LiveGameAlertsEnabled
    {
        get => Bool("live_game_alerts_enabled", true);
        set => Set("live_game_alerts_enabled", value);
    }
    public bool RedZoneAlertsEnabled
    {
        get => Bool("redzone_alerts_enabled", true);
        set => Set("redzone_alerts_enabled", value);
    }
    public bool LowLatencyMode
    {
        get => Bool("low_latency_mode", true);
        set => Set("low_latency_mode", value);
    }
    public bool AudioNormalizationEnabled
    {
        get => Bool("audio_normalization_enabled", true);
        set => Set("audio_normalization_enabled", value);
    }
    public bool AdaptiveQualityEnabled
    {
        get => Bool("adaptive_quality_enabled", true);
        set => Set("adaptive_quality_enabled", value);
    }
    public bool ReducedMotion
    {
        get => Bool("reduced_motion", false);
        set => Set("reduced_motion", value);
    }
    public bool LargeText
    {
        get => Bool("large_text", false);
        set => Set("large_text", value);
    }
    public List<FavoriteTeam> FavoriteTeamProfiles
    {
        get
        {
            try
            {
                if (_prefs.TryGetValue("favorite_team_profiles_v2", out var el) && el.ValueKind == JsonValueKind.Array)
                    return el.Deserialize<List<FavoriteTeam>>() ?? [];
            }
            catch { }
            return [];
        }
        set
        {
            var deduped = value.DistinctBy(t => t.Key).ToList();
            Set("favorite_team_profiles_v2", deduped);
        }
    }
    public List<string> SportsOrder
    {
        get
        {
            var raw = Str("sports_order");
            if (string.IsNullOrEmpty(raw)) return [.. DefaultSportsOrder];
            var saved = raw.Split(',').Select(s => s.Trim()).Where(s => s.Length > 0).ToList();
            return [.. saved, .. DefaultSportsOrder.Where(d => !saved.Contains(d))];
        }
        set => Set("sports_order", string.Join(",", value));
    }

    public bool HasCredentials => SetupComplete || PortalUrl.Length > 0 || StremioAddonUrls.Count > 0;

    public bool ToggleFavoriteTeam(FavoriteTeam team)
    {
        var current = FavoriteTeamProfiles;
        var existing = current.FindIndex(t => t.Key == team.Key);
        if (existing >= 0) { current.RemoveAt(existing); FavoriteTeamProfiles = current; return false; }
        current.Add(team);
        FavoriteTeamProfiles = current;
        return true;
    }

    public bool IsFavoriteTeam(string id, string league) =>
        FavoriteTeamProfiles.Any(t => t.Id == id && t.League.Equals(league, StringComparison.OrdinalIgnoreCase));

    public StreamHealth GetStreamHealth(string target)
    {
        try
        {
            if (_prefs.TryGetValue("stream_health_" + HealthKey(target), out var el))
                return el.Deserialize<StreamHealth>() ?? new StreamHealth();
        }
        catch { }
        return new StreamHealth();
    }
    public void RecordStreamSuccess(string target, long startupMs)
    {
        var h = GetStreamHealth(target);
        var successes = Math.Min(100, h.Successes + 1);
        var avg = successes <= 1 ? startupMs : (h.AverageStartupMs * (successes - 1) + startupMs) / successes;
        SaveHealth(target, h with { Successes = successes, AverageStartupMs = avg, LastUpdatedMs = NowMs() });
    }
    public void RecordStreamFailure(string target)
    {
        var h = GetStreamHealth(target);
        SaveHealth(target, h with { Failures = Math.Min(100, h.Failures + 1), LastUpdatedMs = NowMs() });
    }
    public void RecordStreamStall(string target)
    {
        var h = GetStreamHealth(target);
        SaveHealth(target, h with { Stalls = Math.Min(200, h.Stalls + 1), LastUpdatedMs = NowMs() });
    }
    private void SaveHealth(string target, StreamHealth h) => Set("stream_health_" + HealthKey(target), h);

    internal static string HealthKey(string target) =>
        Convert.ToHexString(SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(target)))[..20].ToLowerInvariant();

    private static long NowMs() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    private string? Str(string key) =>
        _prefs.TryGetValue(key, out var el) && el.ValueKind == JsonValueKind.String ? el.GetString() : null;

    private List<string>? Arr(string key)
    {
        try
        {
            if (_prefs.TryGetValue(key, out var el) && el.ValueKind == JsonValueKind.Array)
                return el.Deserialize<List<string>>();
        }
        catch { }
        return null;
    }

    private bool Bool(string key, bool fallback) =>
        _prefs.TryGetValue(key, out var el) && el.ValueKind is JsonValueKind.True or JsonValueKind.False ? el.GetBoolean() : fallback;

    private void Set(string key, object? value)
    {
        _prefs[key] = JsonSerializer.SerializeToElement(value);
        Save();
    }

    private string Secret(string key) => _secrets.TryGetValue(key, out var v) ? v : "";

    private void SetSecret(string key, string value)
    {
        _secrets[key] = value;
        SaveSecrets();
    }

    private void Load()
    {
        try
        {
            if (File.Exists(_prefsPath))
                _prefs = JsonSerializer.Deserialize<Dictionary<string, JsonElement>>(File.ReadAllText(_prefsPath)) ?? new();
        }
        catch { _prefs = new(); }
        try
        {
            if (File.Exists(_secretsPath))
            {
                var raw = File.ReadAllBytes(_secretsPath);
                var json = OperatingSystem.IsWindows()
                    ? System.Text.Encoding.UTF8.GetString(ProtectedData.Unprotect(raw, null, DataProtectionScope.CurrentUser))
                    : System.Text.Encoding.UTF8.GetString(raw);
                _secrets = JsonSerializer.Deserialize<Dictionary<string, string>>(json) ?? new();
            }
        }
        catch { _secrets = new(); }
    }

    private void Save()
    {
        try { File.WriteAllText(_prefsPath, JsonSerializer.Serialize(_prefs)); } catch { }
    }

    private void SaveSecrets()
    {
        try
        {
            var raw = System.Text.Encoding.UTF8.GetBytes(JsonSerializer.Serialize(_secrets));
            var out_ = OperatingSystem.IsWindows() ? ProtectedData.Protect(raw, null, DataProtectionScope.CurrentUser) : raw;
            File.WriteAllBytes(_secretsPath, out_);
        }
        catch { }
    }
}

File: appleApp/Sources/RallyCore/SettingsStore.swift
import CryptoKit
import Foundation
import Security

/// 1:1 port of `PortalUrlNormalizer` — one source of truth for user-entered URLs.
public enum UrlNormalizer {
    public static func normalizePortal(_ raw: String) -> String {
        var value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty { return "" }
        if !value.lowercased().hasPrefix("http://") && !value.lowercased().hasPrefix("https://") {
            value = "http://" + value
        }
        value = repairRemoteColonTypo(value)
        if value.hasSuffix("/") { value = String(value.dropLast()) }
        for suffix in ["/server/load.php", "/load.php"] where value.hasSuffix(suffix) {
            value = String(value.dropLast(suffix.count))
        }
        value = value.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard var comps = URLComponents(string: value), comps.host != nil else { return "" }
        comps.path = comps.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        comps.path = comps.path.isEmpty ? "" : "/" + comps.path
        return comps.string ?? ""
    }

    public static func normalizeAddon(_ raw: String) -> String? {
        var value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty { return nil }
        if !value.lowercased().hasPrefix("http://") && !value.lowercased().hasPrefix("https://") {
            value = "https://" + value
        }
        guard var comps = URLComponents(string: value), comps.host != nil else { return nil }
        if !comps.path.hasSuffix("manifest.json") {
            comps.path = "/" + (comps.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/manifest.json")
                .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        }
        return comps.string
    }

    public static func normalizeXtreamServer(_ raw: String) -> String {
        var value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty { return "" }
        if !value.lowercased().hasPrefix("http://") && !value.lowercased().hasPrefix("https://") {
            value = "http://" + value
        }
        for suffix in ["/player_api.php", "/get.php"] where value.hasSuffix(suffix) {
            value = String(value.dropLast(suffix.count))
        }
        value = value.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard var comps = URLComponents(string: value), comps.host != nil else { return "" }
        comps.query = nil; comps.fragment = nil
        return comps.string ?? ""
    }

    private static func repairRemoteColonTypo(_ value: String) -> String {
        guard let schemeEnd = value.range(of: "://") else { return value }
        let authStart = schemeEnd.upperBound
        let pathStart = value[authStart...].firstIndex(of: "/") ?? value.endIndex
        var authority = String(value[authStart..<pathStart])
        if authority.hasPrefix("[") || authority.contains("@") { return value }
        guard let colon = authority.lastIndex(of: ":"), colon != authority.startIndex else { return value }
        let suffix = String(authority[authority.index(after: colon)...])
        if suffix.isEmpty || suffix.allSatisfy({ $0.isNumber }) { return value }
        authority.replaceSubrange(colon...colon, with: ".")
        return String(value[..<authStart]) + authority + String(value[pathStart...])
    }
}

public struct StreamHealth: Codable, Sendable, Equatable {
    public var successes: Int
    public var failures: Int
    public var stalls: Int
    public var averageStartupMs: Int64
    public var lastUpdatedMs: Int64
    public init(successes: Int = 0, failures: Int = 0, stalls: Int = 0, averageStartupMs: Int64 = 0, lastUpdatedMs: Int64 = 0) {
        self.successes = successes; self.failures = failures; self.stalls = stalls
        self.averageStartupMs = averageStartupMs; self.lastUpdatedMs = lastUpdatedMs
    }
    /// 1:1 with `PreferencesManager.StreamHealth.score`.
    public var score: Int {
        min(120, max(-240, successes * 24 - failures * 55 - stalls * 12 - Int(averageStartupMs / 750)))
    }
}

/// Settings + secrets. Mirrors `PreferencesManager` keys/defaults.
/// Plist-safe prefs in UserDefaults (stored @Published); token + Xtream
/// password in Keychain.
@MainActor
public final class SettingsStore: ObservableObject {
    public static let defaultAddon = "https://sports.highfly.to/manifest.json"
    public static let defaultSportsOrder = ["NFL", "NCAAF", "NBA", "NCAAB", "MLB", "NHL", "EPL", "La Liga", "Champions League", "Serie A"]

    private let defaults: UserDefaults
    private static let keychainService = "com.shiv.rally.macos"

    @Published public var iptvProvider: IptvProvider {
        didSet { defaults.set(iptvProvider.rawValue, forKey: "iptv_provider") }
    }
    @Published public var portalUrl: String {
        didSet {
            let norm = UrlNormalizer.normalizePortal(portalUrl)
            defaults.set(norm, forKey: "portal_url")
            if norm != portalUrl { portalUrl = norm }
        }
    }
    @Published public var xtreamServerUrl: String {
        didSet {
            let norm = UrlNormalizer.normalizeXtreamServer(xtreamServerUrl)
            defaults.set(norm, forKey: "xtream_server_url")
            if norm != xtreamServerUrl { xtreamServerUrl = norm }
        }
    }
    @Published public var xtreamUsername: String {
        didSet { defaults.set(xtreamUsername.trimmingCharacters(in: .whitespacesAndNewlines), forKey: "xtream_username") }
    }
    @Published public var serialNumber: String {
        didSet { defaults.set(serialNumber, forKey: "serial_number") }
    }
    @Published public var deviceId: String {
        didSet { defaults.set(deviceId, forKey: "device_id") }
    }
    @Published public var stremioAddonUrls: [String] {
        didSet {
            let clean = Array(Set(stremioAddonUrls.compactMap(UrlNormalizer.normalizeAddon))).sorted()
            defaults.set(try? JSONEncoder().encode(clean), forKey: "stremio_addon_urls_json")
            if clean != stremioAddonUrls { stremioAddonUrls = clean }
        }
    }
    @Published public var enabledLeagues: Set<String> {
        didSet { defaults.set(Array(enabledLeagues), forKey: "enabled_leagues") }
    }
    @Published public var favoriteTeams: Set<String> {
        didSet { defaults.set(Array(favoriteTeams), forKey: "favorite_teams") }
    }
    @Published public var favoriteTeamProfiles: [FavoriteTeam] {
        didSet {
            var seen = Set<String>()
            let deduped = favoriteTeamProfiles.filter { seen.insert($0.key).inserted }
            defaults.set(try? JSONEncoder().encode(deduped), forKey: "favorite_team_profiles_v2")
            if deduped != favoriteTeamProfiles { favoriteTeamProfiles = deduped }
        }
    }
    @Published public var liveGameAlertsEnabled: Bool {
        didSet { defaults.set(liveGameAlertsEnabled, forKey: "live_game_alerts_enabled") }
    }
    @Published public var redZoneAlertsEnabled: Bool {
        didSet { defaults.set(redZoneAlertsEnabled, forKey: "redzone_alerts_enabled") }
    }
    @Published public var lowLatencyMode: Bool {
        didSet { defaults.set(lowLatencyMode, forKey: "low_latency_mode") }
    }
    @Published public var audioNormalizationEnabled: Bool {
        didSet { defaults.set(audioNormalizationEnabled, forKey: "audio_normalization_enabled") }
    }
    @Published public var adaptiveQualityEnabled: Bool {
        didSet { defaults.set(adaptiveQualityEnabled, forKey: "adaptive_quality_enabled") }
    }
    @Published public var reducedMotion: Bool {
        didSet { defaults.set(reducedMotion, forKey: "reduced_motion") }
    }
    @Published public var largeText: Bool {
        didSet { defaults.set(largeText, forKey: "large_text") }
    }
    @Published public var setupComplete: Bool {
        didSet { defaults.set(setupComplete, forKey: "setup_complete") }
    }
    @Published public var channelCacheIdentity: String {
        didSet { defaults.set(channelCacheIdentity, forKey: "channel_cache_identity") }
    }
    @Published public var sportsOrder: [String] {
        didSet { defaults.set(sportsOrder.joined(separator: ","), forKey: "sports_order") }
    }

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        iptvProvider = IptvProvider(rawValue: defaults.string(forKey: "iptv_provider") ?? "") ?? .stalker
        portalUrl = defaults.string(forKey: "portal_url") ?? ""
        xtreamServerUrl = defaults.string(forKey: "xtream_server_url") ?? ""
        xtreamUsername = defaults.string(forKey: "xtream_username") ?? ""
        serialNumber = defaults.string(forKey: "serial_number") ?? ""
        deviceId = defaults.string(forKey: "device_id") ?? ""
        if let data = defaults.data(forKey: "stremio_addon_urls_json"),
           let list = try? JSONDecoder().decode([String].self, from: data), !list.isEmpty {
            stremioAddonUrls = list
        } else if let legacy = defaults.string(forKey: "stremio_addon_url")?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !legacy.isEmpty, let norm = UrlNormalizer.normalizeAddon(legacy) {
            stremioAddonUrls = [norm]
        } else {
            stremioAddonUrls = [Self.defaultAddon]
        }
        enabledLeagues = Set(defaults.stringArray(forKey: "enabled_leagues") ?? [])
        favoriteTeams = Set(defaults.stringArray(forKey: "favorite_teams") ?? [])
        if let data = defaults.data(forKey: "favorite_team_profiles_v2"),
           let list = try? JSONDecoder().decode([FavoriteTeam].self, from: data) {
            favoriteTeamProfiles = list
        } else {
            favoriteTeamProfiles = []
        }
        liveGameAlertsEnabled = defaults.object(forKey: "live_game_alerts_enabled") as? Bool ?? true
        redZoneAlertsEnabled = defaults.object(forKey: "redzone_alerts_enabled") as? Bool ?? true
        lowLatencyMode = defaults.object(forKey: "low_latency_mode") as? Bool ?? true
        audioNormalizationEnabled = defaults.object(forKey: "audio_normalization_enabled") as? Bool ?? true
        adaptiveQualityEnabled = defaults.object(forKey: "adaptive_quality_enabled") as? Bool ?? true
        reducedMotion = defaults.bool(forKey: "reduced_motion")
        largeText = defaults.bool(forKey: "large_text")
        setupComplete = defaults.bool(forKey: "setup_complete")
        channelCacheIdentity = defaults.string(forKey: "channel_cache_identity") ?? ""
        if let raw = defaults.string(forKey: "sports_order"), !raw.isEmpty {
            let saved = raw.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
            sportsOrder = saved + Self.defaultSportsOrder.filter { !saved.contains($0) }
        } else {
            sportsOrder = Self.defaultSportsOrder
        }
    }

    public var xtreamPassword: String {
        get { keychainGet("xtream_password") ?? "" }
        set { keychainSet(newValue, account: "xtream_password"); objectWillChange.send() }
    }
    public var authToken: String {
        get { keychainGet("auth_token") ?? "" }
        set { keychainSet(newValue, account: "auth_token"); objectWillChange.send() }
    }
    /// Auto-provisioned once, like Android (`00:1A:79:*`).
    public var macAddress: String {
        get {
            let saved = (defaults.string(forKey: "mac_address") ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if !saved.isEmpty { return saved }
            let mac = "00:1A:79:" + (0..<3).map { _ in String(format: "%02X", UInt8.random(in: 0...255)) }.joined(separator: ":")
            defaults.set(mac, forKey: "mac_address")
            return mac
        }
        set { defaults.set(newValue.trimmingCharacters(in: .whitespacesAndNewlines), forKey: "mac_address"); objectWillChange.send() }
    }

    public var hasCredentials: Bool { setupComplete || !portalUrl.isEmpty || !stremioAddonUrls.isEmpty }

    /// Credentials the user actually entered. The bundled default addon is
    /// excluded: a fresh install must hit onboarding (Android ships no default
    /// addon at all and contacts nothing until configured).
    public var hasUserCredentials: Bool {
        setupComplete || !portalUrl.isEmpty || !xtreamServerUrl.isEmpty || !xtreamUsername.isEmpty
            || stremioAddonUrls.contains { $0.lowercased() != Self.defaultAddon.lowercased() }
    }

    /// First-run gate (Android `startDest`): onboarding until setup is saved
    /// or the user already configured a source.
    public var needsOnboarding: Bool { !setupComplete && !hasUserCredentials }

    /// Toggles by `league:id`, mirroring `toggleFavoriteTeam` (also keeps the legacy name set).
    @discardableResult
    public func toggleFavoriteTeam(_ team: FavoriteTeam) -> Bool {
        if let i = favoriteTeamProfiles.firstIndex(where: { $0.key == team.key }) {
            favoriteTeamProfiles.remove(at: i)
            favoriteTeams.remove(team.name)
            return false
        }
        favoriteTeamProfiles.append(team)
        favoriteTeams.insert(team.name)
        return true
    }
    public func isFavoriteTeam(id: String, league: String) -> Bool {
        favoriteTeamProfiles.contains { $0.id == id && $0.league.caseInsensitiveCompare(league) == .orderedSame }
    }

    public func streamHealth(target: String) -> StreamHealth {
        guard let data = defaults.data(forKey: "stream_health_\(healthKey(target))"),
              let h = try? JSONDecoder().decode(StreamHealth.self, from: data) else { return StreamHealth() }
        return h
    }
    public func recordStreamSuccess(target: String, startupMs: Int64) {
        var h = streamHealth(target: target)
        h.successes = min(100, h.successes + 1)
        h.averageStartupMs = h.successes <= 1 ? startupMs : (h.averageStartupMs * Int64(h.successes - 1) + startupMs) / Int64(h.successes)
        h.lastUpdatedMs = Int64(Date().timeIntervalSince1970 * 1000)
        saveStreamHealth(target: target, health: h)
    }
    public func recordStreamFailure(target: String) {
        var h = streamHealth(target: target)
        h.failures = min(100, h.failures + 1)
        h.lastUpdatedMs = Int64(Date().timeIntervalSince1970 * 1000)
        saveStreamHealth(target: target, health: h)
    }
    public func recordStreamStall(target: String) {
        var h = streamHealth(target: target)
        h.stalls = min(200, h.stalls + 1)
        h.lastUpdatedMs = Int64(Date().timeIntervalSince1970 * 1000)
        saveStreamHealth(target: target, health: h)
    }
    private func saveStreamHealth(target: String, health: StreamHealth) {
        defaults.set(try? JSONEncoder().encode(health), forKey: "stream_health_\(healthKey(target))")
    }
    func healthKey(_ target: String) -> String {
        SHA256.hash(data: Data(target.utf8)).prefix(10).map { String(format: "%02x", $0) }.joined()
    }

    public func clearCredentials() {
        for key in defaults.dictionaryRepresentation().keys { defaults.removeObject(forKey: key) }
        keychainDelete("auth_token"); keychainDelete("xtream_password")
        objectWillChange.send()
    }

    // MARK: - Keychain
    private func keychainGet(_ account: String) -> String? {
        let q: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.keychainService, kSecAttrAccount as String: account,
            kSecReturnData as String: true, kSecMatchLimit as String: kSecMatchLimitOne]
        var out: CFTypeRef?
        guard SecItemCopyMatching(q as CFDictionary, &out) == errSecSuccess,
              let data = out as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }
    private func keychainSet(_ value: String, account: String) {
        let data = Data(value.utf8)
        let q: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.keychainService, kSecAttrAccount as String: account]
        if SecItemCopyMatching(q as CFDictionary, nil) == errSecSuccess {
            SecItemUpdate(q as CFDictionary, [kSecValueData as String: data] as CFDictionary)
        } else {
            var add = q; add[kSecValueData as String] = data
            SecItemAdd(add as CFDictionary, nil)
        }
    }
    private func keychainDelete(_ account: String) {
        let q: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Self.keychainService, kSecAttrAccount as String: account]
        SecItemDelete(q as CFDictionary)
    }
}
